src/builder.ts

```diff
@@ -1330,33 +1345,63 @@ export function buildTypes<
   schemaBuilder?: SchemaBuilder
 ): BuildTypes<TypeMapDefs> {
   const builder = schemaBuilder || new SchemaBuilder(config);
-  addTypes(builder, types);
+  const plugins = config.plugins || [];
+  const blocks = unwrapBlocks(types, builder);
+  const pluggedBlocks = plugins
+    .filter((p) => p.onBeforeBuild)
+    .reduce((blocksAcc, p) => p.onBeforeBuild!(blocksAcc), blocks);
+  pluggedBlocks.forEach(builder.addType);
   return builder.getFinalTypeMap();
 }
 
-function addTypes(builder: SchemaBuilder, types: any) {
-  if (!types) {
-    return;
+function unwrapBlocks(blocks: any, builder: SchemaBuilder): BlockDef[] {
```

Comment:
Separating unwrap from builder so that plugins have chance to cut in and do work without having to unwrap themselves.

Reply:
I think you might be missing what I meant in the other comment above. The version you pushed still doesn't do the same thing as `builder.walkTypes()`, so it would not be useful to call `onBeforeBuild` at this stage because it can't know what all of the types until they are fully unwrapped.

I appreciate the desire to make things more pure from a data perspective, but ultimately I think it will be simpler due to the self-referential property of the types to keep all of this internal in `builder`, as long as we assume that the builder is effectively a protocol for adding the types and calling `getFinalTypeMap` reifies the provided state into a shape that can be provided to the schema.

It might be better to expose what we want to be available from the builder via an explicit interface, rather than the `builder` itsself, sort of what the definition blocks do in the `walkInputType`, `walkOutputType`:

```ts
onBeforeBuild({
  hasType(typeName: string): boolean
  addType(blockDef: BlockDef): void
})
```

Reply:
Yeah these changes came while your first round of feedback dropped and I just completed the iteration anyways. Point received!